### PR TITLE
Add Yandere

### PR DESCRIPTION
**Pull request by LezaiYa**

### CODE REVIEW

File: Patches/CheckGameEndPatch.cs [hunk 1]
```diff
@@ -576,6 +583,11 @@ public bool CheckGameEndByLivingPlayers(out GameOverReason reason)
                     reason = GameOverReason.HumansByVote;
                     winner = CustomWinner.Crewmate;
                 }
+                else if (Jackal < Si && Pel < Si && Imp < Si && BK < Si && Gam < Si && Crew < Si && CM < Si) //胜利
+                {
+                    reason = GameOverReason.ImpostorByKill;
+                    CustomWinnerHolder.ResetAndSetWinner(CustomWinner.Yandere);
```
Comment (Tommy-XL) [inline Patches/CheckGameEndPatch.cs:589]: It seems the victory condition was not done correctly, look at what the victory condition looks like for other NKs and do the same

Comment (LezaiYa) [reply Patches/CheckGameEndPatch.cs:589]: That should do it.

File: main.cs [hunk 2]
```diff
@@ -139,6 +139,8 @@ public class Main : BasePlugin
     public static List<byte> TasklessCrewmate = new();
     public static List<byte> BoobyTrapBody = new();
     public static List<byte> BoobyTrapKiller = new();
+    public static List<byte> NeedKillYandere = new();
+    public static List<byte> ForYandere = new();
```
Comment (Tommy-XL) [inline main.cs:143]: This code can be added to Yandere.cs instead of main

File: Patches/onGameStartedPatch.cs [hunk 3]
```diff
@@ -76,6 +76,8 @@ public static void Postfix(AmongUsClient __instance)
             Main.MedusaBodies = new();
             Main.InfectedBodies = new();
             Main.VirusNotify = new();
+            Main.ForYandere = new();
+            Main.NeedKillYandere = new();
```
Comment (Tommy-XL) [inline Patches/onGameStartedPatch.cs:80]: Also this can be moved in `Yandere.Init();`

File: Roles/Neutral/Yandere.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+                    {
+                        if (yanderes.Is(CustomRoles.Yandere))
+                        {
+                            var playerId = yanderes.PlayerId;
+                            NameColorManager.Add(yanderes.PlayerId,pcs.PlayerId, "#FF0000");
+                            var targetId = pcs.PlayerId;
+                            Target[playerId] = targetId;
+                            TargetArrow.Add(playerId, targetId);
+                        }
+                    }
```
Comment (Gurge44) [inline Roles/Neutral/Yandere.cs:160]: Perfect example of how to write laggy code. 3 foreach cycles within each other put in FixedUpdatePatch. Is there really not a better way to do this?
If not, at least convert them into for cycles using the .ElementAt(i) method if you want better performance

Comment (Tommy-XL) [reply Roles/Neutral/Yandere.cs:160]: Damn, I didn't even notice that there are 3 foreach loops💀

Comment (ryuk201198) [reply Roles/Neutral/Yandere.cs:160]: 💀

File: Patches/MeetingHudPatch.cs [hunk 5]
```diff
@@ -1013,6 +1013,8 @@ public static void Postfix(MeetingHud __instance)
                 case CustomRoles.Lawyer:
                  //   sb.Append(Lawyer.TargetMark(seer, target));
                     break;
+                case CustomRoles.Yandere:
+                    break;
              //   case CustomRoles.Jackal:
```
Comment (Tommy-XL) [inline Patches/MeetingHudPatch.cs:1018]: This code is empty

File: Patches/MeetingHudPatch.cs [hunk 6]
```diff
@@ -1188,6 +1190,10 @@ public static void Postfix(MeetingHud __instance)
             sb.Append(Romantic.TargetMark(seer, target));
             sb.Append(Lawyer.LawyerMark(seer, target));
 
+            //病娇提示
+            if (seer.Is(CustomRoles.Yandere))
+            sb.Append(Yandere.GetTargetArrow(seer));
+
             //会議画面ではインポスター自身の名前にSnitchマークはつけません。
```
Comment (Tommy-XL) [inline Patches/MeetingHudPatch.cs:1197]: Move this code into 

`case CustomRoles.Yandere:
      break;`

Listed above this post

File: Patches/PlayerControlPatch.cs [hunk 7]
```diff
@@ -2771,6 +2782,22 @@ public static void Postfix(PlayerControl __instance)
...
+                Wildling.OnFixedUpdate(player);
+                Spiritcaller.OnFixedUpdate(player);
+                Pitfall.OnFixedUpdate(player);
+                Swooper.OnFixedUpdate(player);
+                Wraith.OnFixedUpdate(player);
+                Chameleon.OnFixedUpdate(player);
+                Spy.OnFixedUpdate(player);
+                Yandere.OnFixedUpdate(player);
+            //    Alchemist.OnFixedUpdate(player);
+
```
Comment (Tommy-XL) [inline Patches/PlayerControlPatch.cs:2800]: This code is already outdated because I rewrote `FixedUpdate`

Comment (LezaiYa) [reply Patches/PlayerControlPatch.cs:2800]: .

File: Patches/onGameStartedPatch.cs (hunk 3 above)
Comment (Tommy-XL) [reply Patches/onGameStartedPatch.cs:79]: Still not done

File: Roles/Neutral/Yandere.cs [hunk 8]
```diff
@@ -0,0 +1,218 @@
...
+                            var targetId = pcs.PlayerId;
+                            Target[playerId] = targetId;
+                            TargetArrow.Add(playerId, targetId);
+                        }
+                    }                                      
+                    break;
+                }
+            }
+        }
+    }
```
Comment (Tommy-XL) [inline Roles/Neutral/Yandere.cs:167]: You just removed the loops, this will probably just cause bugs

**Comment by LezaiYa** [comment]
I'll close this PR and sync the new commit before pulling it again ._.

**Pull request closed without merging**